sdk/nodejs/resource.ts

```diff
@@ -384,26 +389,57 @@ export class Output<T> {
+            // not known itself, then the result we return should not be known.
+            const resultIsKnown = Promise.all([isKnown, innerIsKnown]).then(([k1, k2]) => k1 && k2);
 
-                const transformed = await func(v);
-                if (Output.isInstance(transformed)) {
-                    // Note: if the func returned a Output, we unwrap that to get the inner value
-                    // returned by that Output.  Note that we are *not* capturing the Resources of
-                    // this inner Output.  That's intentional.  As the Output returned is only
-                    // supposed to be related this *this* Output object, those resources should
-                    // already be in our transitively reachable resource graph.
-                    return await transformed.promise();
-                } else {
-                    return transformed;
+            return new Output<U>(resources, promise.then(async v => {
+                try {
+                    // During previews do not perform the apply if the engine was not able to
+                    // give us an actual value for this Output.
+                    const perform = await isKnown;
+                    const isDryRun = testingOptions.isDryRun || runtime.isDryRun();
+
+                    if (isDryRun && !perform) {
+                        // We couldn't run the function, our new Output is definitely **not** known.
+                        innerIsKnownResolve(false);
+                        return <U><any>undefined;
+                    }
+
+                    const transformed = await func(v);
+                    if (Output.isInstance(transformed)) {
+                        // Note: if the func returned a Output, we unwrap that to get the inner value
+                        // returned by that Output.  Note that we are *not* capturing the Resources of
+                        // this inner Output.  That's intentional.  As the Output returned is only
+                        // supposed to be related this *this* Output object, those resources should
+                        // already be in our transitively reachable resource graph.
+
+                        // The callback func has produced an inner Output that may be 'known' or 'unknown'.
+                        // We have to properly forward that along to our outer output.  That way the Outer
+                        // output doesn't consider itself 'known' then the inner Output did not.
+                        transformed.isKnown.then(innerIsKnownResolve, innerIsKnownReject);
+                        return await transformed.promise();
+                    } else {
+                        // We successfully ran the inner function.  Our new Output should be considered known.
+                        innerIsKnownResolve(true);
+                        return transformed;
+                    }
+                } catch (err) {
```

Comment:
There’s a subtle case for this catch where the `await transformer.promise()` could throw after we’ve already resolved this promise. Not sure precise promise semantics in this case - but potentially a concern?

Reply:
I think the rejection stuff actually wrong.  i think we shoudl only ever resolve the inner promise, not ever reject it.

Reply:
(i'm not sure it would matter, but i thin kit will simplify things greatly).